File: relay/earthlysearch/grammar.go
package earthlysearch

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// The Earthly NIP-50 extension grammar (docs/GEO_SEARCH_REWRITE.md §4).
//
// A search string is free text plus space-separated `key:value` tokens.
// Only tokens with a KNOWN key are consumed as extensions; anything else
// (including URLs and unknown keys) stays in the free text, so foreign
// search strings pass through untouched. The same grammar is serialized by
// the client facade in src/lib/search/ — both sides are pinned by the golden
// vectors in spec/search-grammar-vectors.json.

const GrammarVersion = 1

// GrammarExtensions is advertised via the /earthly-search capability
// endpoint (referenced from NIP-11) for client feature detection.
var GrammarExtensions = []string{
	"bbox", "point", "rel", "near", "radius",
	"label", "tag", "ref", "start-after", "start-before", "sort",
}

const (
	RelIntersects = "intersects"
	RelContains   = "contains"
	RelWithin     = "within"

	SortRelevance = "relevance"
	SortDistance  = "distance"
	SortRecent    = "recent"
	SortScale     = "scale"
)

// SearchParams is the parsed form of an extended NIP-50 search string.
type SearchParams struct {
	Text string `json:"text"`

	BBox     *[4]float64 `json:"bbox"`  // w,s,e,n
	Point    *[2]float64 `json:"point"` // lon,lat
	Rel      string      `json:"rel"`   // geo relation of indexed geometry to query shape
	Near     string      `json:"near"`  // geohash proximity origin
	RadiusKm float64     `json:"radiusKm"`

	Labels   []string `json:"labels"`
	Hashtags []string `json:"hashtags"`
	Refs     []string `json:"refs"`

	StartAfter  int64 `json:"startAfter"`  // NIP-52 start >= (epoch seconds)
	StartBefore int64 `json:"startBefore"` // NIP-52 start <= (epoch seconds)

	Sort string `json:"sort"`
}

// HasGeo reports whether any spatial constraint is present.
func (p SearchParams) HasGeo() bool {
	return p.BBox != nil || p.Point != nil || p.Near != ""
}

func parseTimeValue(v string) (int64, error) {
	if ts, err := strconv.ParseInt(v, 10, 64); err == nil {
		return ts, nil
	}
	t, err := time.Parse("2006-01-02", v)
	if err != nil {
		return 0, fmt.Errorf("expected epoch seconds or YYYY-MM-DD, got %q", v)
	}
	return t.Unix(), nil
}

func parseRadius(v string) (float64, error) {
	var factor float64
	var num string
	switch {
	case strings.HasSuffix(v, "km"):
		factor, num = 1, strings.TrimSuffix(v, "km")
	case strings.HasSuffix(v, "m"):
		factor, num = 0.001, strings.TrimSuffix(v, "m")
	default:
		return 0, fmt.Errorf("radius needs a km or m suffix, got %q", v)
	}
	f, err := strconv.ParseFloat(num, 64)
	if err != nil || f <= 0 {
		return 0, fmt.Errorf("invalid radius %q", v)
	}
	return f * factor, nil
}

func isValidGeohash(v string) bool {
	if len(v) < 1 || len(v) > 9 {
		return false
	}
	for _, c := range v {
		if !strings.ContainsRune(geohashBase32, c) {
			return false
		}
	}
	return true
}

// ParseSearch splits an extended search string into free text and typed
// parameters. A recognized key with a malformed value is a hard error (the
// relay rejects the filter with a useful message) — silently dropping a geo
// constraint would return wrong results.
func ParseSearch(s string) (SearchParams, error) {
	params := SearchParams{Rel: RelIntersects, Sort: SortRelevance}
	var text []string

	for _, token := range strings.Fields(s) {
		key, value, found := strings.Cut(token, ":")
		if !found || value == "" {
			text = append(text, token)
			continue
		}

		switch key {
		case "bbox":
			box, err := parseBboxTag(value)
			if err != nil {
				return params, fmt.Errorf("bbox: %w", err)
			}
			params.BBox = box

		case "point":
			parts := strings.Split(value, ",")
			if len(parts) != 2 {
				return params, fmt.Errorf("point needs lon,lat, got %q", value)
			}
			lon, err1 := strconv.ParseFloat(parts[0], 64)
			lat, err2 := strconv.ParseFloat(parts[1], 64)
			if err1 != nil || err2 != nil || !validLonLat(lon, lat) {
				return params, fmt.Errorf("invalid point %q", value)
			}
			params.Point = &[2]float64{lon, lat}

		case "rel":
			switch value {
			case RelIntersects, RelContains, RelWithin:
				params.Rel = value
			default:
				return params, fmt.Errorf("rel must be intersects|contains|within, got %q", value)
			}

		case "near":
			if !isValidGeohash(value) {
				return params, fmt.Errorf("near expects a geohash, got %q", value)
			}
			params.Near = strings.ToLower(value)

		case "radius":
			km, err := parseRadius(value)
			if err != nil {
				return params, err
			}
			params.RadiusKm = km

		case "label":
			params.Labels = append(params.Labels, value)

		case "tag":
			params.Hashtags = append(params.Hashtags, value)

		case "ref":
			params.Refs = append(params.Refs, value)

		case "start-after":
			ts, err := parseTimeValue(value)
			if err != nil {
				return params, fmt.Errorf("start-after: %w", err)
			}
			params.StartAfter = ts

		case "start-before":
			ts, err := parseTimeValue(value)
			if err != nil {
				return params, fmt.Errorf("start-before: %w", err)
			}
			params.StartBefore = ts

		case "sort":
			switch value {
			case SortRelevance, SortDistance, SortRecent, SortScale:
				params.Sort = value
			default:
				return params, fmt.Errorf("sort must be relevance|distance|recent|scale, got %q", value)
			}

		default:
			// Unknown key — leave it in the free text (URLs, "re:", foreign
			// relay extensions). Only known keys are grammar.
			text = append(text, token)
		}
	}

	if params.Sort == SortDistance && params.Point == nil && params.Near == "" {
		return params, fmt.Errorf("sort:distance requires point: or near:")
	}
	if params.RadiusKm > 0 && params.Near == "" && params.Point == nil {
		return params, fmt.Errorf("radius: requires near: or point:")
	}

	params.Text = strings.Join(text, " ")
	return params, nil
}

File: src/lib/search/grammar.ts
import type { SearchQuery } from './types'

/**
 * Serialize a SearchQuery to the extended NIP-50 search string.
 *
 * Canonical token order (pinned by spec/search-grammar-vectors.json):
 * text, bbox, point, rel, near, radius, label*, tag*, ref*, start-after,
 * start-before, sort. Defaults (rel:intersects, sort:relevance) are omitted.
 *
 * Foreign relays see unknown tokens as ignorable NIP-50 extensions; use
 * `stripExtensions` when the target relay is known not to support them.
 */
export function buildSearchString(query: SearchQuery): string {
	const tokens: string[] = []

	const text = query.text?.trim()
	if (text) tokens.push(text)

	if (query.bbox) {
		validateBbox(query.bbox)
		tokens.push(`bbox:${query.bbox.join(',')}`)
	}

	if (query.point) {
		const [lon, lat] = query.point
		if (!isValidLonLat(lon, lat)) {
			throw new Error(`invalid point: ${query.point.join(',')}`)
		}
		tokens.push(`point:${lon},${lat}`)
	}

	if (query.rel && query.rel !== 'intersects') {
		tokens.push(`rel:${query.rel}`)
	}

	if (query.near) {
		if (!isValidGeohash(query.near)) {
			throw new Error(`invalid geohash for near: ${query.near}`)
		}
		tokens.push(`near:${query.near.toLowerCase()}`)
	}

	if (query.radiusKm !== undefined && query.radiusKm > 0) {
		if (!query.near && !query.point) {
			throw new Error('radiusKm requires near or point')
		}
		tokens.push(`radius:${formatRadius(query.radiusKm)}`)
	}

	for (const label of query.labels ?? []) {
		tokens.push(`label:${requireTokenSafe(label, 'label')}`)
	}
	for (const hashtag of query.hashtags ?? []) {
		tokens.push(`tag:${requireTokenSafe(hashtag, 'hashtag')}`)
	}
	for (const ref of query.refs ?? []) {
		tokens.push(`ref:${requireTokenSafe(ref, 'ref')}`)
	}

	if (query.startAfter !== undefined) {
		tokens.push(`start-after:${formatTime(query.startAfter)}`)
	}
	if (query.startBefore !== undefined) {
		tokens.push(`start-before:${formatTime(query.startBefore)}`)
	}

	if (query.sort && query.sort !== 'relevance') {
		if (query.sort === 'distance' && !query.near && !query.point) {
			throw new Error('sort:distance requires near or point')
		}
		tokens.push(`sort:${query.sort}`)
	}

	return tokens.join(' ')
}

/**
 * Reduce a query to plain NIP-50 text for relays without the Earthly
 * extension (feature-detect via fetchSearchCapability).
 */
export function stripExtensions(query: SearchQuery): string {
	return query.text?.trim() ?? ''
}

/** True when the query has any constraint beyond free text. */
export function hasExtensions(query: SearchQuery): boolean {
	return buildSearchString(query) !== (query.text?.trim() ?? '')
}

const GEOHASH_BASE32 = '0123456789bcdefghjkmnpqrstuvwxyz'

function isValidGeohash(value: string): boolean {
	if (value.length < 1 || value.length > 9) return false
	return [...value.toLowerCase()].every((c) => GEOHASH_BASE32.includes(c))
}

function isValidLonLat(lon: number, lat: number): boolean {
	return (
		Number.isFinite(lon) &&
		Number.isFinite(lat) &&
		lon >= -180 &&
		lon <= 180 &&
		lat >= -90 &&
		lat <= 90
	)
}

function validateBbox(bbox: [number, number, number, number]): void {
	const [w, s, e, n] = bbox
	if (!isValidLonLat(w, s) || !isValidLonLat(e, n) || w > e || s > n) {
		throw new Error(`invalid bbox: ${bbox.join(',')}`)
	}
}

function formatRadius(km: number): string {
	if (km < 1) return `${Math.round(km * 1000)}m`
	return `${km}km`
}

function formatTime(value: number | string): string {
	if (typeof value === 'number') {
		if (!Number.isInteger(value) || value <= 0) {
			throw new Error(`invalid epoch time: ${value}`)
		}
		return String(value)
	}
	if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) {
		throw new Error(`invalid date (want YYYY-MM-DD or epoch seconds): ${value}`)
	}
	return value
}

/** Grammar values must not contain whitespace (they are space-delimited tokens). */
function requireTokenSafe(value: string, kind: string): string {
	if (!value || /\s/.test(value)) {
		throw new Error(`invalid ${kind} value: ${JSON.stringify(value)}`)
	}
	return value
}
